codegen/ace/src/circuit.rs

```diff
@@ -0,0 +1,487 @@
+                        match op {
+                            Value(_) => {}
+                            Add(l, r) | Sub(l, r) | Mul(l, r) => {
+                                if !self.visited.contains(&r) {
+                                    self.visited.insert(r);
+                                    self.process.push(r)
+                                };
+                                if !self.visited.contains(&l) {
+                                    self.visited.insert(l);
+                                    self.process.push(l)
+                                };
+                            }
+                        };
+                    }
+                }
+            }
+        }
+        None
+    }
+}
+
+/// Remap the children of an operation. Typically when migrating an operation
+/// between data structures. Any index withtin the inputs is left untouched.
+pub fn remap_op(
+    n_inputs: usize,
+    op: Operation,
+    mapping: &BTreeMap<NodeIndex, NodeIndex>,
+) -> Operation {
+    let remap = |i: NodeIndex| {
+        if i < n_inputs.into() {
+            i
+        } else {
+            mapping[&i]
+        }
+    };
+    match op {
+        Operation::Add(l, r) => Operation::Add(remap(l), remap(r)),
+        Operation::Sub(l, r) => Operation::Sub(remap(l), remap(r)),
+        Operation::Mul(l, r) => Operation::Mul(remap(l), remap(r)),
+        Operation::Value(Value::Constant(_)) => op,
+        _ => unreachable!(""),
+    }
+}
+
+#[derive(Default)]
+pub struct CircuitBuilder {
+    // The ACE circuit being built
+    pub circuit: Circuit,
+    // A cache of nodes already inserted in the circuit, used to avoid duplicates
+    cache: BTreeMap<Operation, NodeIndex>,
+    // Constants zero and one are inserted first so that they can be used as
+    // identity and absorbing element in addition and multiplication.
+    pub zero: NodeIndex,
+    pub one: NodeIndex,
+    // Stores the position of the alpha input once it's set.
+    pub alpha: Option<NodeIndex>,
+    // Tracks the next power of alpha to be used. Starts at 1=alpha^0.
+    next_alpha: NodeIndex,
+}
+
+impl CircuitBuilder {
+    pub fn new(n_inputs: usize) -> (Self, Vec<NodeIndex>) {
+        let mut cache: BTreeMap<Operation, NodeIndex> = BTreeMap::new();
+        let mut circuit = Circuit::new(n_inputs);
+        let [zero, one] = [0, 1].map(|c| {
+            let op = Operation::Value(Value::Constant(c as u64));
+            let out = circuit.push(op);
+            cache.insert(op, out);
+            out
+        });
+        (
+            Self {
+                cache,
+                circuit,
+                zero,
+                one,
+                alpha: None,
+                next_alpha: one,
+            },
+            (0..n_inputs).map(NodeIndex::from).collect(),
+        )
+    }
+
+    // Append an operation to the circuit. A new operation is added only if
+    // syntactically different from any operations already added to the circuit
```

Comment:
nit: "... if it is syntactically different from any operation already added ..."